### DISCUSSION TITLE
Issue with container image package identification with manifest annotations

### DISCUSSION THREAD

**Discussion by aarongoldenthal** [renovatebot/renovate#31347] [Request Help] [answered]
### What would you like help with?

Other

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab, Renovate v38.76.1 (`renovate-runner` v18.44.0)

### Please tell us more about your question or problem

I was troubleshooting an issue with Renovate _incorrectly_ identifying container image packages, and think I understand what's it's doing, but wondering if there's a way to work around it. It appears this may not be a change in Renovate behavior, but a change in Docker hub official image manifests.

Previously (it changed somewhere between May 7 and 23), Renovate would correctly identify publicly available container images from other registries and pull release notes, for example:

![image](https://github.com/user-attachments/assets/96e73b50-5fd5-4ada-8d58-e6c83ee8c785)

Sometime in mid May (between May 7 and 23), that behavior seemed to change and Renovate started identifying those images as their base images for links in a PR, for example:

![image](https://github.com/user-attachments/assets/1b42a7e9-2b2f-4411-acef-bd68f5c6397e)

The new versions are being correctly identified, but the links in the PR are to the base image repositories (in this example the `package` link goes to https://hub.docker.com/_/golang and the `source` link goes to https://github.com/docker-library/golang), and the release notes are no longer being pulled.

This has been seen with several images, all of which have Docker hub official images as their base.

After a little investigation, it appears that Docker hub started adding `annotations` to their manifests, which Renovate is now using to flag the base image based on `org.opencontainers.image.source` and `org.opencontainers.image.url` (at least that's the only reference I can find to those repositories). For example:

```ps
PS C:\> docker manifest inspect registry.gitlab.com/gitlab-ci-utils/container-images/go-test:2.4.0
{
    "schemaVersion": 2,
    "mediaType": "application/vnd.oci.image.manifest.v1+json",
    "config": {
        "mediaType": "application/vnd.oci.image.config.v1+json",
        "size": 2857,
        "digest": "sha256:c488f401d3e77d3e4cdb7de0800d27df344d8f4c6d7fff61b78c02bd470a3339"
    },
    "layers": [
        ...
    ],
    "annotations": {
            "com.docker.official-images.bashbrew.arch": "amd64",
            "org.opencontainers.image.base.digest": "sha256:216266c86fc4dcef5619930bd394245824c2af52fd21ba7c6fa0e618657d4c3b",
            "org.opencontainers.image.base.name": "alpine:3.20",
            "org.opencontainers.image.created": "2024-05-22T12:22:57Z",
            "org.opencontainers.image.revision": "cf7a37dedf1fd5a25ca72075645368d1e3c30c4a",
            "org.opencontainers.image.source": "https://github.com/docker-library/golang.git#cf7a37dedf1fd5a25ca72075645368d1e3c30c4a:1.22/alpine3.20",
            "org.opencontainers.image.url": "https://hub.docker.com/_/golang",
            "org.opencontainers.image.version": "1.22.3-alpine3.20"
    }
}
```

My images (e.g. the `registry.gitlab.com/gitlab-ci-utils/container-images/go-test` image) are built with `kaniko`, which doesn't add `annotations` (I have yet to even find a way for it to add them), so the base `annotations` cascade through to the final image.

Ideally, the solution is to update the `annotations` to be correct for the final image, but while sorting that out with `kaniko` I'm wondering if this is expected Renovate behavior and if there's a way to point renovate to the correct package and source repositories for the final image?

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by aarongoldenthal** [comment]
Thinking more about the timing, it may not have been a change in Docker hub, but more likely the GitLab 17.0 container registry changes. In any case, the images pulled from the GitLab container registry had `annotations` after that point.

**Reply by viceice** [reply, marked as answer]
yes, that's expected behavior. you should also override the docker labels.

**Reply by aarongoldenthal** [reply]
Thanks @viceice. I am already setting the applicable `LABEL`s, but unfortunately they just set the image config and don't override the manifest annotations.

**Reply by aarongoldenthal** [reply]
For reference to anyone else who ends up here, see https://docs.renovatebot.com/modules/datasource/docker/#description
